File: pSCAN-refactor/DisjointSet.cpp
//
// Created by yche on 8/6/17.
//

#include "DisjointSet.h"

void DisjointSet::MakeDisjointSetBatch(ui capacity) {
    parent = vector<int>(capacity);
    rank = vector<int>(capacity);
    for (ui i = 0; i < capacity; i++) { parent[i] = i; }
    std::fill(rank.begin(), rank.end(), 0);
}

DisjointSet::DisjointSet(ui capacity) {
    MakeDisjointSetBatch(capacity);
}

int DisjointSet::FindRoot(int u) {
    int x = u;
    while (parent[x] != x) { x = parent[x]; }

    while (parent[u] != x) {
        // heuristic: path compression
        int tmp = parent[u];
        parent[u] = x;
        u = tmp;
    }
    return x;
}

bool DisjointSet::IsSameSet(int u, int v) {
    return FindRoot(u) == FindRoot(v);
}

void DisjointSet::Union(int u, int v) {
    int ru = FindRoot(u);
    int rv = FindRoot(v);

    if (ru == rv) { return; }

    // heuristic: union by rank
    // Link: ru and rv
    if (rank[ru] < rank[rv]) {
        parent[ru] = rv;
    } else if (rank[ru] > rank[rv]) {
        parent[rv] = ru;
    } else {
        parent[rv] = ru;
        ++rank[ru];
    }
}

File: python_experiments/algorithm_vis/disjoint_set.py
class DisjointSet:
    def __init__(self, capacity):
        self.parent = range(capacity)
        self.rank = [0] * capacity

    def find_root(self, u):
        x = u
        while self.parent[x] != x:
            x = self.parent[x]

        # heuristics: path compression
        while self.parent[u] != x:
            tmp = self.parent[u]
            self.parent[u] = x
            u = tmp
        return x

    def union(self, u, v):
        ru = self.find_root(u)
        rv = self.find_root(v)
        if ru == rv:
            return

        # heuristics: union by rank
        if self.rank[ru] < self.rank[rv]:
            self.parent[ru] = rv
        elif self.rank[ru] > self.rank[rv]:
            self.parent[rv] = ru
        else:
            self.parent[rv] = ru
            self.rank[ru] += 1
